### PR TITLE
Enable pre-build for napa.js

### PR DESCRIPTION
**Pull request by fs-eire**
## Onboard pre-build for napa.js.

* Consume package **node-pre-gyp** and **node-pre-gyp-github**
* introduce a script (scripts/install.js) for `npm install`
  - default behavior is to try fetch the pre-build binaries. If failed, fall back to build.
  - support flag `--no-fetch` in `npm install` to skip downloading the pre-build binaries
  - support flag `--no-build` in `npm install` to skip building from sources
  - If both flag specified, the script do nothing. Napa.js will not work in this state. (This is used for embeded purpose, which will use a different build apporach)
  - In NPM <= 4.x, the script will run `npm run prepare` because npm won't run it after `npm install`.
* Update the building for the example module.
  - Build napa binary, napa binding and an example module (for test) together
  - Binplace to <root>/bin/
  - update the corresponding test files

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -9,18 +9,28 @@
...
+    "mocha": "^3.5.0",
+    "typescript": "^2.4.2",
+    "@types/node": "^8.0.22",
+    "@types/mocha": "^2.2.41",
+    "markdown-table": "^1.1.1"
   },
   "scripts": {
     "benchmark": "node benchmark/bench.js",
-    "install": "cmake-js compile",
+    "install": "node scripts/install.js",
```
Comment (asibross) [inline package.json]: >node  [](start = 16, length = 5)

I am not sure this is needed. Please double check.

File: scripts/install.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+// ==== Try to fetch the pre-build binaries ====
+if (!skipFetch) {
+	try {
+		log.info('NAPA_INSTALL', 'trying to download the pre-build binaries.');
+		execSync(fetchCommand, { 'stdio': 'inherit' });
+
+		log.info('NAPA_INSTALL', 'done successfully by download.');
+		skipBuild = true;
+	} catch (e) {
+		errorCode = e.status;
```
Comment (asibross) [inline scripts/install.js]: >errorCode = e.status; [](start = 2, length = 21)

do you expect to continue the script execution if an error occurred?
If not then return e.status.

File: scripts/install.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+			errorCode = e.status;
+		}
+	}
+}
+
+if (errorCode != 0) {
+	log.error('NAPA_INSTALL', 'failed to install napajs.');
+}
+
+process.exit(errorCode);
```
Comment (asibross) [inline scripts/install.js]: Is there a reason you don't use return here?

File: test/module-test.ts [hunk 4]
```diff
@@ -36,7 +36,7 @@ describe('napajs/module', function () {
         it('napa module', () => {
             return napaZone.execute((thisFilePath: string) => {
                 var assert = require("assert");
-                var napaModule = require(thisFilePath + '/module/addon/build/simple-napa-addon.napa');
+                var napaModule = require(thisFilePath + '/../bin/simple-napa-addon.napa');
```
Comment (asibross) [inline test/module-test.ts]: >napa- [](start = 72, length = 5)

drop the napa, just sample-addon.napa

File: scripts/install.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+var log = require('npmlog');
+var execSync = require('child_process').execSync;
+
+// Default command
+var fetchCommand = 'node-pre-gyp install --fallback-to-build=false';
+var buildCommand = 'cmake-js compile';
+
+// ==== Determine install options ====
+// Skip the fetch stage. '--no-fetch', or '--fetch=false'
+var skipFetch = process.env.hasOwnProperty('npm_config_fetch') && !process.env['npm_config_fetch'];
```
Comment (asibross) [inline scripts/install.js]: >'npm_config_fetch' [](start = 43, length = 18)

Please explain where the transformation between --no-fetch and napa_config_fetch happens.

**Comment by asibross** [comment]
> set(NAPA_ROOT ${CMAKE_CURRENT_SOURCE_DIR}/../../..)

Since this is now being built from the root cmake file. You can just use the ${PROJECT_SOURCE_DIR} macro,

---
Refers to: test/module/addon/CMakeLists.txt:5 in f40b1ae. [](commit_id = f40b1ae3bc62b9e82a90982fce2a40acf70c6b3c, deletion_comment = False)

**Comment by asibross** [comment]
:clock1:

File: package.json (hunk 1 above)
Comment (fs-eire) [inline package.json]: It's simply not working (at least) under Windows.

---
In reply to: [133244153](https://github.com/Microsoft/napajs/pull/41#discussion_r133244153) [](ancestors = 133244153)

Comment (fs-eire) [inline package.json]: Error msg -

> scripts/install.js

'scripts' is not recognized as an internal or external command,
operable program or batch file.

---
In reply to: [133266454](https://github.com/Microsoft/napajs/pull/41#discussion_r133266454) [](ancestors = 133266454,133244153)

File: scripts/install.js (hunk 2 above)
Comment (fs-eire) [inline scripts/install.js]: Yes, we have a fall-back logic by default. If it fails to fetch the binaries it builds from sources

---
In reply to: [133245349](https://github.com/Microsoft/napajs/pull/41#discussion_r133245349) [](ancestors = 133245349)

File: scripts/install.js (hunk 3 above)
Comment (fs-eire) [inline scripts/install.js]: return a non-zero value here will not set the process exitcode.

I used that approach and npm always think the operation is completed successfully no matter what return value.

---
In reply to: [133245757](https://github.com/Microsoft/napajs/pull/41#discussion_r133245757) [](ancestors = 133245757)

File: package.json [hunk 6]
```diff
@@ -9,18 +9,28 @@
     "type": "git",
     "url": "https://github.com/Microsoft/napajs.git"
   },
+  "binary": {
+    "module_name": "napa-binding",
```
Comment (daiyip) [inline package.json]: Does this include all files in `bin` directory? in that case I suggest to use `napa-bin`

**Comment by fs-eire** [comment]
> set(NAPA_ROOT ${CMAKE_CURRENT_SOURCE_DIR}/../../..)

It is not true. We have the 'npm run retest' script to build the test only. I think this is helpful for developers to update and debug tests.

---
In reply to: [322529425](https://github.com/Microsoft/napajs/pull/41#issuecomment-322529425) [](ancestors = 322529425)

---
Refers to: test/module/addon/CMakeLists.txt:5 in f40b1ae. [](commit_id = f40b1ae3bc62b9e82a90982fce2a40acf70c6b3c, deletion_comment = False)

File: scripts/install.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+// Skip the build stage. '--no-build', or '--build=false'
+var skipBuild = process.env.hasOwnProperty('npm_config_build') && !process.env['npm_config_build'];
+
+// Use debug build
+if (process.env.hasOwnProperty('npm_config_debug') && process.env['npm_config_debug']) {
```
Comment (daiyip) [inline scripts/install.js]: Please change all TABs into 4 spaces.

File: scripts/install.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+if (errorCode == 0) {
+	var npmVersion = execSync('npm --version').toString().trim();
+	var npmMajorVersion = npmVersion.split('.')[0];
+	var npmMajorVersionNumber = parseInt(npmMajorVersion);
+	if (npmMajorVersionNumber < 4) {
+		// Before npm 4.x, the 'prepare' script will not run automatically by npm.
+		// We have to run it explicitly in this script.
+
+		try {
+			log.info('NAPA_INSTALL', `Current NPM version=${npmVersion}. NPM below 4.x does not recognize script 'prepare'.`);
```
Comment (daiyip) [inline scripts/install.js]: Have consistent style in log. (Like use upper-case letter or lower-case for starting alphabet)

File: scripts/install.js (hunk 5 above)
Comment (fs-eire) [inline scripts/install.js]: From what I observed and learned from other project (such as node-pre-gyp), NPM parses package.json and commandline arguments and generates some env variables with prefix npm_config_

However, actually I didn't find any info about this on npm documentation. The most related would be this:
[https://docs.npmjs.com/misc/config#environment-variables](https://docs.npmjs.com/misc/config#environment-variables)


---
In reply to: [133246502](https://github.com/Microsoft/napajs/pull/41#discussion_r133246502) [](ancestors = 133246502)

File: package.json (hunk 6 above)
Comment (fs-eire) [inline package.json]: the 'module-name' field is used exactly for node-pre-gyp searching for file <module-name>.node

---
In reply to: [133269664](https://github.com/Microsoft/napajs/pull/41#discussion_r133269664) [](ancestors = 133269664)

File: test/module-test.ts (hunk 4 above)
Comment (fs-eire) [inline test/module-test.ts]: Why do we call it a 'sample'?

---
In reply to: [133246133](https://github.com/Microsoft/napajs/pull/41#discussion_r133246133) [](ancestors = 133246133)

Comment (daiyip) [inline test/module-test.ts]: test-addon.napa

File: package.json (hunk 1 above)
Comment (asibross) [inline package.json]: weird, npm documentation specifies that you can.
[https://docs.npmjs.com/misc/scripts#examples](https://docs.npmjs.com/misc/scripts#examples)


---
In reply to: [133266593](https://github.com/Microsoft/napajs/pull/41#discussion_r133266593) [](ancestors = 133266593,133266454,133244153)

Comment (asibross) [inline package.json]: It's not an issue though.

---
In reply to: [133492526](https://github.com/Microsoft/napajs/pull/41#discussion_r133492526) [](ancestors = 133492526,133266593,133266454,133244153)

File: scripts/install.js (hunk 2 above)
Comment (asibross) [inline scripts/install.js]: Then please add some comment about it.

---
In reply to: [133266822](https://github.com/Microsoft/napajs/pull/41#discussion_r133266822) [](ancestors = 133266822,133245349)

File: test/module-test.ts (hunk 4 above)
Comment (asibross) [inline test/module-test.ts]: sorry, I meant simple-addon.napa

---
In reply to: [133275193](https://github.com/Microsoft/napajs/pull/41#discussion_r133275193) [](ancestors = 133275193,133246133)

**Comment by asibross** [comment]
> set(NAPA_ROOT ${CMAKE_CURRENT_SOURCE_DIR}/../../..)

ok.
However, doesn't specifying the project('simple-napa-addon') interfere with the napa project?

---
In reply to: [322553298](https://github.com/Microsoft/napajs/pull/41#issuecomment-322553298) [](ancestors = 322553298,322529425)

---
Refers to: test/module/addon/CMakeLists.txt:5 in f40b1ae. [](commit_id = f40b1ae3bc62b9e82a90982fce2a40acf70c6b3c, deletion_comment = False)

**Comment by fs-eire** [comment]
> set(NAPA_ROOT ${CMAKE_CURRENT_SOURCE_DIR}/../../..)

CMake supports nested projects. Our case is not the typical usage, but I think it is generally OK.

---
In reply to: [322821763](https://github.com/Microsoft/napajs/pull/41#issuecomment-322821763) [](ancestors = 322821763,322553298,322529425)

---
Refers to: test/module/addon/CMakeLists.txt:5 in f40b1ae. [](commit_id = f40b1ae3bc62b9e82a90982fce2a40acf70c6b3c, deletion_comment = False)

File: package.json [hunk 9]
```diff
@@ -9,18 +9,28 @@
...
+    "module_path": "./bin/",
+    "host": "https://github.com/Microsoft/napajs/releases/download/",
+    "remote_path": "{version}"
+  },
+  "dependencies": {
+    "node-pre-gyp": "^0.6.36",
+    "npmlog": "^4.1.2"
```
Comment (asibross) [inline package.json]: >"npmlog": "^4.1.2" [](start = 4, length = 18)

Is this a must dependency?

**Pull request merged**